test/cpp/api/any.cpp

```diff
@@ -0,0 +1,200 @@
+#include <catch.hpp>
+
+#include <torch/torch.h>
+
+#include <torch/detail/any.h>
+#include <torch/nn/modules/any.h>
+
+#include <algorithm>
+#include <string>
+
+using namespace torch;
+using namespace torch::nn;
+using namespace torch::detail;
+
+using Catch::StartsWith;
+
+TEST_CASE("any") {
+  SECTION("gets the correct value for the right type") {
+    SECTION("int") {
+      auto any = make_any(5);
+      // const and non-const types have the same typeid()
+      REQUIRE(any.try_get<int>() != nullptr);
+      REQUIRE(any.try_get<const int>() != nullptr);
+      REQUIRE(any.get<int>() == 5);
+    }
+    SECTION("const int") {
+      auto any = make_any(5);
+      REQUIRE(any.try_get<const int>() != nullptr);
+      REQUIRE(any.try_get<int>() != nullptr);
+      REQUIRE(any.get<const int>() == 5);
+    }
+    SECTION("const char*") {
+      auto any = make_any("hello");
+      REQUIRE(any.try_get<const char*>() != nullptr);
+      REQUIRE(any.get<const char*>() == std::string("hello"));
+    }
+    SECTION("std::string") {
+      auto any = make_any(std::string("hello"));
+      REQUIRE(any.try_get<std::string>() != nullptr);
+      REQUIRE(any.get<std::string>() == "hello");
+    }
+    SECTION("pointers") {
+      std::string s("hello");
+      std::string* p = &s;
+      auto any = make_any(p);
+      REQUIRE(any.try_get<std::string*>() != nullptr);
+      REQUIRE(*any.get<std::string*>() == "hello");
+    }
+    SECTION("references") {
+      std::string s("hello");
+      const std::string& t = s;
+      auto any = make_any(t);
+      REQUIRE(any.try_get<std::string>() != nullptr);
+      REQUIRE(any.get<std::string>() == "hello");
+    }
+  }
+  SECTION("move is allowed") {
+    auto any = make_any(5);
+    SECTION("construction") {
+      auto copy = make_any(std::move(any));
+      REQUIRE(copy.try_get<int>() != nullptr);
+      REQUIRE(copy.get<int>() == 5);
+    }
+    SECTION("assignment") {
+      auto copy = make_any(10);
+      copy = std::move(any);
+      REQUIRE(copy.try_get<int>() != nullptr);
+      REQUIRE(copy.get<int>() == 5);
+    }
+  }
+  SECTION("type_info is correct") {
+    SECTION("int") {
+      auto any = make_any(5);
+      REQUIRE(any.type_info().hash_code() == typeid(int).hash_code());
```

Comment:
Can't you just do a direct `==` here? Why is the `hash_code` necessary

Reply:
> There is no guarantee that the same std::type_info instance will be referred to by all evaluations of the typeid expression on the same type, although std::type_info::hash_code of those type_info objects would be identical, as would be their std::type_index.